src/Databases/PostgreSQL/DatabaseMaterializedPostgreSQL.cpp

```diff
@@ -164,8 +256,116 @@ void DatabaseMaterializedPostgreSQL::createTable(ContextPtr local_context, const
         return;
     }
 
-    throw Exception(ErrorCodes::NOT_IMPLEMENTED,
-        "Create table query allowed only for ReplacingMergeTree engine and from synchronization thread");
+    const auto & create = query->as<ASTCreateQuery>();
+    if (!create->attach)
+        throw Exception(ErrorCodes::QUERY_NOT_ALLOWED, "CREATE TABLE is not allowed for database engine {}. Use ATTACH TABLE instead", getEngineName());
+
+    /// Create ReplacingMergeTree table.
+    auto query_copy = query->clone();
+    auto * create_query = assert_cast<ASTCreateQuery *>(query_copy.get());
+    create_query->attach = false;
+    create_query->attach_short_syntax = false;
+    DatabaseAtomic::createTable(StorageMaterializedPostgreSQL::makeNestedTableContext(local_context), table_name, table, query_copy);
+
+    /// Attach MaterializedPostgreSQL table.
+    attachTable(table_name, table, {});
+}
+
+
+void DatabaseMaterializedPostgreSQL::attachTable(const String & table_name, const StoragePtr & table, const String & relative_table_path)
+{
+    if (CurrentThread::isInitialized() && CurrentThread::get().getQueryContext())
```

Comment:
Why so strange condition in `attachTable` method? Why the behavior depends on caller thread?